homeassistant/components/recorder/__init__.py

```diff
@@ -191,38 +199,25 @@ def run(self):
                 self.queue.task_done()
                 return
 
-            elif event.event_type == EVENT_TIME_CHANGED:
+            if event.event_type == EVENT_TIME_CHANGED:
                 self.queue.task_done()
                 continue
 
             session = Session()
             dbevent = Events.from_event(event)
             session.add(dbevent)
-
-            for _ in range(0, RETRIES):
-                try:
-                    session.commit()
-                    break
-                except sqlalchemy.exc.OperationalError as e:
-                    log_error(e, retry_wait=QUERY_RETRY_WAIT,
-                              rollback=True)
+            self._commit(session)
 
             if event.event_type != EVENT_STATE_CHANGED:
                 self.queue.task_done()
                 continue
 
+            # Can possibly be combined into single session with the above
             session = Session()
             dbstate = States.from_event(event)
-
-            for _ in range(0, RETRIES):
-                try:
-                    dbstate.event_id = dbevent.event_id
-                    session.add(dbstate)
-                    session.commit()
-                    break
-                except sqlalchemy.exc.OperationalError as e:
-                    log_error(e, retry_wait=QUERY_RETRY_WAIT,
-                              rollback=True)
+            dbstate.event_id = dbevent.event_id
+            session.add(dbstate)
```

Comment:
As mentioned in the PR level comment, SQLAlchemy requires the work to be resubmitted on failures.  I understand the desire to remove duplicated code.  Perhaps a function that takes a callable, pseudo code:

```
    def do_work():
        dbstate = States.from_event(event)
        dbstate.event_id = dbevent.event_id
        session.add(dbstate)
    retry_helper(do_work)

Then have a retry_helper that retries work and performs the commit, etc.

Not sure that's ideal, but it would work reliably, rather than wasting cycles resubmitting failed transactions.